app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.ts

```diff
@@ -0,0 +1,262 @@
+  // Specifications. Calling the PermissionController with them will
+  // cause an error to be thrown. Instead, we will use the approval result
+  // from the ApprovalController to form a CAIP-25 permission later.
+  let legacyApproval;
+  const haveLegacyPermissions =
+    Object.keys(legacyRequestedPermissions).length > 0;
+  if (haveLegacyPermissions) {
+    if (!legacyRequestedPermissions[RestrictedMethods.eth_accounts]) {
+      legacyRequestedPermissions[RestrictedMethods.eth_accounts] = {};
+    }
+
+    if (!legacyRequestedPermissions[PermissionNames.permittedChains]) {
+      legacyRequestedPermissions[PermissionNames.permittedChains] = {};
+    }
+
+    if (isSnapId(origin)) {
+      delete legacyRequestedPermissions[PermissionNames.permittedChains];
+    }
+
+    legacyApproval = await requestPermissionApprovalForOrigin(
+      legacyRequestedPermissions,
+    );
+  }
+
+  let grantedPermissions: GrantedPermissions = {};
+  // Request permissions from the PermissionController for any permissions other
+  // than eth_accounts and permittedChains in the params. If no permissions
+  // are in the params, then request empty permissions from the PermissionController
+  // to get an appropriate error to be returned to the dapp.
+  if (
+    (Object.keys(requestedPermissions).length === 0 &&
+      !haveLegacyPermissions) ||
+    Object.keys(requestedPermissions).length > 0
+  ) {
+    const [frozenGrantedPermissions] = await requestPermissionsForOrigin(
+      requestedPermissions,
+    );
+    // permissions are frozen and must be cloned before modified
+    grantedPermissions = { ...frozenGrantedPermissions };
+  }
+
+  if (legacyApproval) {
+    let newCaveatValue = {
+      requiredScopes: {},
+      optionalScopes: {},
+      isMultichainOrigin: false,
+    };
+
+    if (isSnapId(origin)) {
+      newCaveatValue.optionalScopes = {
+        'wallet:eip155': {
+          accounts: [],
+        },
+      };
+    } else {
+      newCaveatValue = setPermittedEthChainIds(
+        newCaveatValue,
+        legacyApproval.approvedChainIds,
+      );
+    }
+
+    newCaveatValue = setEthAccounts(
+      newCaveatValue,
+      legacyApproval.approvedAccounts,
+    );
+
+    const permissions = getPermissionsForOrigin() || {};
+    let caip25Endowment = permissions[Caip25EndowmentPermissionName];
+    const existingCaveatValue = caip25Endowment?.caveats?.find(
+      ({ type }) => type === Caip25CaveatType,
+    )?.value as Caip25CaveatValue | undefined;
+    if (existingCaveatValue) {
+      if (existingCaveatValue.isMultichainOrigin) {
+        return end(
+          new Error(
+            'Cannot modify permission granted via the Multichain API. Either modify the permission using the Multichain API or revoke permissions and request again.',
+          ),
+        );
+      }
+
+      updateCaveat(
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
+        newCaveatValue,
+      );
```

Comment:
my brain might have been jello when i wrote this. I don't know why i decided it made sense to update the caveat of an existing CAIP-25 permission rather than overwrite the existing one entirely...

Reply:
🤔 The alternative would be just to use `grantPermissions` in both cases?

Reply:
Is there a functional difference?

Reply:
functionally, not really. but this is a regression in behavior vs main

Reply:
removed here https://github.com/MetaMask/metamask-extension/pull/29213